imports/lib/models/Model.ts

```diff
@@ -0,0 +1,187 @@
+
+class Model<T extends z.ZodObject<any, 'strict'>> {
+  name: string;
+
+  schema: T;
+
+  collection: Mongo.Collection<z.output<T>>;
+
+  constructor(name: string, schema: T) {
+    this.name = name;
+    this.schema = schema;
+    this.collection = new Mongo.Collection(name);
+    AllModels.add(this);
+  }
+
+  async insertAsync(doc: z.input<T>): Promise<z.output<T>['_id']> {
+    const cleaned: z.output<T> = await IsInsert.withValue(true, async () => {
+      return this.schema.parseAsync(doc);
+    });
+    return this.collection.insert(cleaned as any) as any;
+  }
+
+  // As a note, we don't accept full documents as modifiers because our
+  // transform functions don't make sense (e.g. we won't accept a `createdAt`
+  // timestamp, but we expect one to be there)
+  async updateAsync(
+    selector: Selector<z.output<T>>,
+    modifier: Exclude<Mongo.Modifier<z.output<T>>, z.output<T>>,
+    options: {
+      multi?: boolean | undefined;
+      upsert?: boolean | undefined;
+      arrayFilters?: { [identifier: string]: any }[] | undefined;
+    } = {},
+  ): Promise<number> {
+    // Types should prevent passing a full document as a modifier, but verify at
+    // runtime to be sure
+    const rawModifier: Mongo.Modifier<z.output<T>> = modifier as any;
+    if (modifierIsWholeDoc(rawModifier)) {
+      throw new Error('Cannot update with a full document');
+    }
+
+    const setOnInsertFields = await IsUpsert.withValue(true, async () => {
+      return optionalPreservingPartial(this.schema).parseAsync({});
+    });
+    const setFields = await IsUpdate.withValue(true, async () => {
+      return optionalPreservingPartial(this.schema).parseAsync({});
+    });
+
+    const mergedModifier: Mongo.Modifier<z.output<T>> = {
+      ...rawModifier,
+      $setOnInsert: {
+        ...rawModifier.$setOnInsert,
+        ...setOnInsertFields,
+      },
+      $set: {
+        ...rawModifier.$set,
+        ...setFields,
+      },
+    };
+
+    return this.collection.update(selector, mergedModifier, options);
+  }
+
+  async upsertAsync(
+    selector: Selector<z.output<T>>,
+    modifier: Exclude<Mongo.Modifier<z.output<T>>, z.output<T>>,
+    options: {
+      multi?: boolean | undefined;
+    } = {},
+  ): Promise<{
+    numberAffected?: number | undefined;
+    insertedId?: z.output<T>['_id'] | undefined;
+  }> {
+    const rawModifier: Mongo.Modifier<T> = modifier as any;
+    if (modifierIsWholeDoc(rawModifier)) {
+      throw new Error('Cannot update with a full document');
+    }
+
+    const setOnInsertFields = await IsUpsert.withValue(true, async () => {
+      return optionalPreservingPartial(this.schema).parseAsync({});
+    });
+    const setFields = await IsUpdate.withValue(true, async () => {
+      return optionalPreservingPartial(this.schema).parseAsync({});
+    });
+
+    const mergedModifier: Mongo.Modifier<z.output<T>> = {
+      ...rawModifier,
+      $setOnInsert: {
+        ...rawModifier.$setOnInsert,
```

Comment:
This appears to have `setOnInsertFields` overriding anything the user put in `$setOnInsert`.  I'd have expected the inverse behavior -- that if the user specifies a `createdAt` that we'd use the explicit value they gave, rather than inferring one.  Should we switch the order here?

Ditto `$set` and `setFields` below.

Reply:
So I actually thought that we had written `{created,updated}{At,By}` to not be overridable by callers, but I just mis-remembered. I think I can make 3 changes to get the right behavior here:

- Define the various custom timestamp fields as `z.date().optional().transform(...)` instead of `z.undefined().optional()` (i.e. allow them to be explicitly passed in).
- Update the transforms to preserve a value if it's already set
- Pass the `$set` and `$setOnInsert` parameters to `update` into `schema.parse`.

Again, I'll write tests to cover all of this but I'll push up the behavior modifications for now.

Reply:
To be fair I also didn't remember the existing behavior, but the types at the `update` callsite permitted me to provide an explicit `updatedAt` field as either Date or undefined, so I expected that if I did provide a value it wouldn't go to `/dev/null`.  I think I don't have an opinion yet on whether we should differentiate between "explicitly provided a value which was undefined" and "did not set that key in the update object", but I think that's already the case with needing to provide `$set` vs `$unset`.  I could probably be convinced that we should forbid explicit values for our autoValue fields if you like that better, but I feel like being able to specify the value if you want a particular one is probably more flexible and just as usable in practice.